### ISSUE TITLE
[Feature]: Set cutoff tokens in SDXL prompt at will using the built-in feature and the "," symbol

### ISSUE THREAD

**Issue by Nekos4Lyfe** [vladmandic/sdnext#2179]
### Feature description

I'm developing an extension and discovered something strange in the configuration with the CLIPTokenizer on Automatic1111 for and SDXL model i dowloaded from civitai.

I was running CLIPtokenizer.encode(text) on a given input string  ' text' the tokenizer seems to append cutoff tokens |start-of-text| and |end-of-text| between every token.

![blub1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/130230016/598edf9c-4962-4958-bec7-f214e1fd03ab)

Here is a a printout of Automatic1111/modules/shared/sd_model.cond_stage_model.wrapped:

'####model.cond_stage_model.embedders.wrapped[0]####'
{'_backward_hooks': OrderedDict(),
 '_backward_pre_hooks': OrderedDict(),
 '_buffers': OrderedDict(),
 '_forward_hooks': OrderedDict(),
 '_forward_hooks_with_kwargs': OrderedDict(),
 '_forward_pre_hooks': OrderedDict(),
 '_forward_pre_hooks_with_kwargs': OrderedDict(),
 '_input_key': 'txt',
 '_is_full_backward_hook': None,
 '_is_trainable': False,
 '_load_state_dict_post_hooks': OrderedDict(),
 '_load_state_dict_pre_hooks': OrderedDict(),
 '_modules': OrderedDict([('transformer',
                           CLIPTextModel(
  (text_model): CLIPTextTransformer(
    (embeddings): CLIPTextEmbeddings(
      (token_embedding): EmbeddingsWithFixes(
        (wrapped): Embedding(49408, 768)
      )
      (position_embedding): Embedding(77, 768)
    )
    (encoder): CLIPEncoder(
      (layers): ModuleList(
        (0-11): 12 x CLIPEncoderLayer(
          (self_attn): CLIPAttention(
            (k_proj): Linear(in_features=768, out_features=768, bias=True)
            (v_proj): Linear(in_features=768, out_features=768, bias=True)
            (q_proj): Linear(in_features=768, out_features=768, bias=True)
            (out_proj): Linear(in_features=768, out_features=768, bias=True)
          )
          (layer_norm1): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
          (mlp): CLIPMLP(
            (activation_fn): QuickGELUActivation()
            (fc1): Linear(in_features=768, out_features=3072, bias=True)
            (fc2): Linear(in_features=3072, out_features=768, bias=True)
          )
          (layer_norm2): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        )
      )
    )
    (final_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
  )
))]),
 '_non_persistent_buffers_set': set(),
 '_parameters': OrderedDict(),
 '_state_dict_hooks': OrderedDict(),
 '_state_dict_pre_hooks': OrderedDict(),
 '_ucg_rate': 0.0,
 'device': 'cuda',
 'layer': 'hidden',
 'layer_idx': 11,
 'legacy_ucg_val': None,
 'max_length': 77,
 'return_pooled': False,

 '**tokenizer': CLIPTokenizer**(name_or_path='openai/clip-vit-large-patch14', vocab_size=49408, model_max_length=77, is_fast=False, padding_side='right', truncation_side='right', special_tokens={'bos_token': AddedToken("<|startoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=True), 'eos_token': AddedToken("<|endoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=True), 'unk_token': AddedToken("<|endoftext|>", rstrip=False, lstrip=False, single_word=False, normalized=True), 'pad_token': '<|endoftext|>'}, clean_up_tokenization_spaces=True),

 'train': <function disabled_train at 0x7a5f1d845870>,
 'training': True}
'####END OF model.cond_stage_model.embedders.wrapped[0]####'

I realize this problem arose because I tokenized the text piece by piece in my code. I expect (or well , I sincerly hope ) that this configuration exists in A1111 because  the tokenizer tokenizes the prompt "all-at-once" rather then "word-by-word" as I do currently. 

Should be noted that would be pretty bad for prompt comprehension if cutoff tokens in A1111 (or SD Next ) were placed as shown in the first image. But I trust that is under control. 

Anyway , I fixed this issue in my extension by filtering out the cutoff tokens from the configured A1111 tokenizer and letting the user assign the symbol "<" and ">" to place a |start-of-text| and |end-of-text| token in the output. This made prompt comprehension much better and was very convenient.

![cutoff1](https://github.com/AUTOMATIC1111/stable-diffusion-webui/assets/130230016/ac019050-b95c-45be-a6e3-7f6753bd33fd)

So, I think this would be a really neat feature to have for SD Next in the prompt box. You could let users write a special symbol in the prompt box like commas "," "," (or something else) and have the tokenizer interpret that as a cutoff token. 

### Proposed workflow

When using SDXL model 
1. Write "," in the prompt to place |start-of-text| token there (token ID# is  49406) (context dependent)
2. Write "," in the prompt to place |end-of-text| token (token ID# is  49407)
(also context dependent)
3. Ensure there is a |start-of-text|  token as the first token in the prompt and a |end-of-text|  as the the last token in the prompt

TLDR:
You want the prompt "A , B C  , D " to become tokenized to 
"< sot > A < eot > < sot > B C < eot > < sot > D < eot > "

where < sot > = |start-of-text| token 
and < eot > = |end-of-text| token

### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits

**Comment by vladmandic** [comment]
i totally agree except why use fixed char (`,`) for that? make it a user configurable string, especially since `,` is often intercepted or parsed by many things.

**Comment by Nekos4Lyfe** [comment]
Sounds even better!

**Comment by vladmandic** [comment]
re: default - yes, there should be a good default, i just don't think `,` is it - its overused by a lot of things, plus ppl place it in prompts randomly without considerations. i'm ok with any not already used char or string.

regarding strength - i feel that too many times "advanced" params are added without understanding their impact. again, totally not against it if value can be measured and there is actual use-case - cannot be just for the sake of "its possible".

**Comment by Nekos4Lyfe** [comment]
That is reasonable. In that case , how about using the dot  `.` symbol?

**Comment by vladmandic** [comment]
Simple and it works. Thumbs up.

**Comment by Nekos4Lyfe** [comment]
@vladmandic  One more thing. Turns out this feature also exists with SD 1.5 .

![Cutoff SD 1 5](https://github.com/vladmandic/automatic/assets/130230016/1834c5f6-8363-431c-b3be-ba149af21e35)

**Comment by vladmandic** [comment]
even better

**Comment by Nekos4Lyfe** [comment]
![psuedocode](https://github.com/vladmandic/automatic/assets/130230016/0a70d87f-24f5-48ea-865d-c0ad95f98c48)

Some additional info which can be useful.  Easy method to implement this feature is to check the IDs from the tokenizer output. SDXL uses a the same tokenizer vocabulary for both 1280 and 768 dimension vectors. Here is an image shown from Automatic1111.

If an ID equals 13 ("." symbol) or 269 (". " symbol) then replace that ID in the ID-list with IDs 49407 ("end-of-text") and 49406 ("start-of-text") in that order. 

Vocabulary for the Tokenizer can be found in tokenizer.json on the huggingface website for that CLIP model. For example OpenCLIP used for the 768 dimension vectors in SDXL : https://huggingface.co/openai/clip-vit-large-patch14/blob/main/tokenizer.json

**Comment by vladmandic** [comment]
just following up, where are we with this?

**Comment by Nekos4Lyfe** [comment]
No change from earlier. I haven't written any code for SD Next if that is what you are wondering. 
Busy with other stuff. How are things on your end?

**Comment by vladmandic** [comment]
still interested in this, but love to see a PR :)

**Comment by vladmandic** [comment]
added, uses explicit `BOS` and `EOS` strings in prompt.
